### ISSUE TITLE
Add spanner support for Set, LinkedList and other collections

### ISSUE THREAD

**Issue by dmitry-s** [spring-cloud/spring-cloud-gcp#967]
Currently spanner supports only arrays and types assignable from ArrrayList. Other collection types should be supported, with optional user provided converters.

**Comment by ChengyuanZhao** [comment]
Sounds like a good idea!

**Comment by ChengyuanZhao** [comment]
Upon closer exam I have some concerns:

Because the client library natively gives or takes arrays and lists, adding support for Set and other collection types (or any arbitrary type really since the user provides his own converter), is just a thin convenience for the user:  we apply this converter automatically for him or her rather than he or she doing it separately. 

However, because the converters must be of unique types (for example from ArrayList to/from Set), this means this single converter will be used in all cases of List columns being converted to Set. However, the behavior for different columns in the same table or different tables may need to be different, but we cannot support that. 

There are 2 issues that come to mind: 

1. The converters should be implementing the same Converter interface that we have used here and in the Datastore module. Because of type erasure, I think we can only distinguish between `Converter<List, Set>` and `Converter<List, OtherType>` and that we cannot distinguish between `Converter<List<Integer>, Set<Integer>>` and `Converter<List<Integer>, Set<String>>.` 
As a result, if we have one column that is an array of numbers and another that is an array of string, we would have a problem.

2. If we have two columns and for one when converting to set we take the entire value of the string, while for the other we may want to calculate some identifying value from the string and use that to determine if it should be in the set. We cannot support this because the type of the converter in the two situations would be the same and we can only have 1 such converter. 

We could restrict the use cases to exclude situations like the 2 mentioned above, but then I think the functionality is severely limited, and the user would be better off and in much better control if he applied the conversions himself (like in a getter or setter in the POJO). 

@dmitry-s @meltsufin what do you think?

**Comment by meltsufin** [comment]
@ChengyuanZhao Do you still have the same concerns after our conversation about this behavior in the Datastore implementation?

**Comment by ChengyuanZhao** [comment]
@meltsufin Yes I think the first point is addressed, though I think the second point still stands if I change the example to use PriorityQueue instead of Set.  

1. Indeed `Converter<List<Integer>, PriorityQueue<Integer>>` and `Converter<List<Integer>, PriorityQueue<Car>>` cannot be distinguished from each other, but that's OK because the converter for List -> PriorityQueue will always do the simple add-each-item and assume the items are all just `Object`. If Car does not implement Comparable, then it will just crash and say so. I think this should be fine, and I think this point is addressed. 

2. In this situation we have 2 columns of the same type (Strings), but how they should be treated by final collection type (PriorityQueue) differs. For example let's say we want LastName column (Strings) to be ascending priority, and StudentId (Strings) prioritized  in some arbitrary way that is different (could be as simple as descending or complicated like seniors alphabetically followed by juniors alphabetically, etc...) . There wouldn't be a way to use 2 different String comparators since they would be using the same converter. 
We could just say "too bad , not supported" to this second case, but I still feel like if we just tell users to do this custom-collection conversion in their own getters and setters then we can avoid this.

**Comment by meltsufin** [comment]
@ChengyuanZhao I'm not really following your second concern. In your example, in order to have a `PriorityQueue<String>` that sorts in descending order, you would need to construct the `PriorityQueue` with a custom comparator. I assume that the conversion service just uses the empty-arg constructor of the collection, where the comparator would be lost anyway.
In principle, converting an array into a priority queue is like unsafe downcasting, where something like priority order customization can be lost. Going from an array to a set is a bit safer because hashcode and equals are part of `Object`.
The point is that it's not 100% safe, but I guess it's worth it to allow the use of various collection types like `Set`, `ArrayList`, `LinkedList`, etc., backed by Cloud Spanner's `ARRAY` column type.

**Comment by ChengyuanZhao** [comment]
@meltsufin Yes I think we are on the same page. I agree with your last sentence:
" The point is that it's not 100% safe, but I guess it's worth it to allow the use of various collection types like Set, ArrayList, LinkedList, etc., backed by Cloud Spanner's ARRAY column type."

It sounds like the point here is to provide convenient conversion to a small set of other collection types that are very similar to Arrays/ArrayLists, and not to provide a mechanism to allow conversion between Arrays/ArrayLists and arbitrary other collection types. 

Would you agree with that? If so, then I think we should hardcode support for this fixed small set of other types (set, linkedlist, immutablelist, etc..) rather than trying to use converters and conversion service. The converters and conversion service is open ended and meant to support as yet undetermined collection conversions, which given this situation doesn't sound like what we intend to support.

**Comment by meltsufin** [comment]
Yes, I think we're on the same page. I think we should get some feedback on this question from the Spring Data team. @olivergierke @mp911de

**Comment by dmitry-s** [comment]
@ChengyuanZhao I think we should give users ability to implement converters for their own collections. For example, if they want to use ImmutableSet, there should be a way to do that, for example using conversion service. I believe that's how it is implemented in mongo spring data.

**Comment by meltsufin** [comment]
@ChengyuanZhao I noticed that [in Objectify](https://github.com/objectify/objectify/wiki/Entities#collections), for example, the `PriorityQueue` with a custom comparator can be achieved by creating the `PriorityQueue` in the class initialization.
```java
class MyEntity {
    PriorityQueue<String> names = new PriorityQueue((a, b) -> -a.compareTo(b));
}
```
However, it doesn't look like something like this is supported in Spring Data.

**Comment by ChengyuanZhao** [comment]
@meltsufin that's a very interesting idea ! Perhaps we can kinda emulate this  ( maybe something like an annotation on the collection property that is given some sort of anonymous function that provides a new instance of the desired custom collection)
